src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1143,6 +1147,7 @@ impl LinkCollector<'_, '_> {
                     None
                 }
             },
+            Res::Primitive(_) => Some(ItemLink { link: ori_link, link_text, did: None, fragment }),
```

Comment:
This won't check that the disambiguator is correct, it unconditionally resolves the link. I would refactor the body of the `Res::Def` arm into a closure (or function) and call that to check this is correct. You can get the `kind` and `id` from the `kind_side_channel` you just updated.

Please also add a test case:
```rust
//! [struct@str::trim] //~ ERROR incompatible link
```

Reply:
That makes sense, I made the refactor. Now I'm getting this error for the mismatched disambiguator:
```
error: unresolved link to `str::trim`
  --> /home/max/documents/projects/rust/src/test/rustdoc/intra-doc/incompatible-primitive-disambiguator.rs:3:6
   |
 3 | //! [struct@str::trim]
   |      ^^^^^^^^^^^^^^^^
   |      |
   |      this link resolves to the associated function `trim`, which is not in the type namespace
   |      help: to link to the associated function, add parentheses: `str::trim()`
```
which isn't the one we want. It seems like it's coming from a `ResolutionFailure::WrongNamespace` and that the disambiguator is never making it to the part of the code I'm touching.

Reply:
I'm not sure without getting a debug build. `WrongNamespace` is very wrong but I'm also baffled how you could end up with one - the parsing code is relatively simple and I would expect it to panic before returning the wrong thing: https://github.com/rust-lang/rust/blob/06b0900a2dc53ed4b5fa98c859fec5c224929eb8/src/librustdoc/passes/collect_intra_doc_links.rs#L1473
Can you try to find how you end up with a `Disambiguator::Namespace`?

Reply:
We don't end up with a namespace disambiguator, the type namespace is implied by the `struct@` disambiguator. Then, in `resolve_with_disambiguator`, we get here:

https://github.com/rust-lang/rust/blob/06b0900a2dc53ed4b5fa98c859fec5c224929eb8/src/librustdoc/passes/collect_intra_doc_links.rs#L1285-L1292

which finds the associated method in the value namespace and yields the namespace error.

Reply:
Hmm, that seems like an existing bug. If you use `struct@Vec::new`, does it have a similar misleading error? If so I think it's ok not to fix it here and mention it in https://github.com/rust-lang/rust/issues/77150; you could just use a different test that works approximately right instead.

Reply:
Yep:
```
warning: unresolved link to `Vec::new`
 --> src/lib.rs:1:6
  |
1 | //! [struct@Vec::new]
  |      ^^^^^^^^^^^^^^^
  |      |
  |      this link resolves to the associated function `new`, which
 is not in the type namespace
  |      help: to link to the associated function, add parentheses:
 `Vec::new()`
  |
  = note: `#[warn(broken_intra_doc_links)]` on by default

warning: 1 warning emitted
```

Reply:
Commented on the issue and pushed an updated test.